contracts/core/OriginCore.sol

```diff
@@ -414,4 +482,119 @@ contract OriginCore is OriginCoreInterface, OriginCoreConfig {
 
         return kernelHash;
     }
+
+    /**
+     * @notice This function checks vote validity.
+     *
+     * @dev This method will revert the transaction if validator is not
+     *       eligible for voting.
+     *
+     * @param _transitionHash The hash of the transition part of the meta-block
+     *                        header at the source block.
+     * @param _voteHash The hash of the vote object.
+     * @param _v V of the signature.
+     * @param _r R of the signature.
+     * @param _s S of the signature.
+     *
+     * @return currentHeight_ Height of current meta-block.
+     * @return signer_ Address of validator who has signed the vote.
+     */
+    function isVoteValid(
+        bytes32 _transitionHash,
+        bytes32 _voteHash,
+        uint8 _v,
+        bytes32 _r,
+        bytes32 _s
+
+    )
+        private
+        view
+        returns (uint256 currentHeight_, address signer_)
+    {
+        // As per https://github.com/ethereum/go-ethereum/pull/2940
+        bytes memory prefix = "\x19Ethereum Signed Message:\n32";
+        _voteHash = keccak256(
+            abi.encodePacked(
+                prefix,
+                _voteHash
+            )
+        );
+
+        signer_ = ecrecover(
+            _voteHash,
+            _v,
+            _r,
+            _s
+        );
+
+        /* Header of last meta block. */
+        MetaBlock.Header storage latestMetaBlockHeader = reportedHeaders[head];
+        /* Current meta-block height. */
+        currentHeight_ = latestMetaBlockHeader.kernel.height.add(1);
+        uint256 weight = stake.weight(currentHeight_, signer_);
+
+        require(
+            weight > 0,
+            "Only validator with non zero weight can vote."
+        );
+
+        require(
+            seals[_transitionHash].validators[signer_] == false,
+            "Vote already verified for this validator."
+        );
+
+    }
+
+    /**
+     * @notice This function saves vote into the seal.
+     *
+     * @dev This method assumes that vote has been successfully verified. It
+     *      will revert the transaction if validator already voted for this
+     *      transition object.
+     *
+     * @param _transitionHash The hash of the transition part of the meta-block
+     *                    header at the source block.
+     * @param _signer Address of validator who has signed vote object.
+     * @param _height Height of current meta-block.
+     */
+    function saveVote(
+        bytes32 _transitionHash,
+        address _signer,
+        uint256 _height
+    )
+        private
+        returns(uint256 totalVoteWeight_)
+    {
+        MetaBlock.Seal storage seal = seals[_transitionHash];
+
+        seal.validators[_signer] = true;
+
+        totalVoteWeight_ = seal.totalVoteWeight.add(stake.weight(_height,_signer));
+
+        seal.totalVoteWeight = totalVoteWeight_;
```

Comment:
should a test cover that this cannot be commented if it is required?

Reply:
This is already covered in test `it should increase total weight on successful verification of vote` 👍

Reply:
Then why did the tests pass when this line was commented? 🤔 Can you please try and make sure that the tests fail if you remove this line?

Reply:
Sarvesh already tested successfully